tensorboard/components/tensor_widget/selection.ts

```diff
@@ -0,0 +1,340 @@
+  private viewDims: number[];
+  private rowStart: number;
+  private colStart: number;
+  private rowCount: number;
+  private colCount: number;
+
+  /**
+   * Constructor of TensorElementSelection
+   * @param shape Shape of the tensor in which the viewing and selection is
+   *   taking place.
+   * @param slicingSpec The current slicing spec for the tensor.
+   * @param rowStart The starting row of selection, within the indices
+   *   framework of the original tensor (i.e., *not* with respect to the
+   *   slicing spec.)
+   * @param colStart The starting column of selection, within the indices
+   *   framework of the original tensor (i.e., *not* with respect to the
+   *   slicing spec.)
+   * @param rowCount How many rows are selected.
+   * @param colCount How many columns are selected.
+   */
+  constructor(
+    private readonly shape: Shape,
+    readonly slicingSpec: TensorViewSlicingSpec,
+    rowStart?: number,
+    colStart?: number,
+    rowCount?: number,
+    colCount?: number
+  ) {
+    if (size(this.shape) === 0) {
+      throw new Error(
+        `TensorElementSelection doesn't support tensor with zero elements.`
+      );
+    }
+
+    for (let i = 0; i < slicingSpec.slicingDimsAndIndices.length; ++i) {
+      this.sliceDims.push(slicingSpec.slicingDimsAndIndices[i].dim);
+      const index = slicingSpec.slicingDimsAndIndices[i].index;
+      if (index === null) {
+        throw new Error(
+          `Failed to create TensorElementSelection due to ` +
+            `undetermined slicing index at dimension ${i}`
+        );
+      }
+      this.sliceIndices.push(index);
+    }
+
+    // Sanity check the size of the the slicing dimensions.
+    if (this.rank() > 0 && this.sliceDims.length >= this.rank()) {
+      throw new Error(
+        `Expected sliceDims to have a length less than rank ${this.rank}, ` +
+          `but got length ${this.sliceDims.length}`
+      );
+    }
+
+    // Determine the viewing dimensions.
+    this.viewDims = [];
+    for (let i = 0; i < this.rank(); ++i) {
+      if (this.sliceDims.indexOf(i) === -1) {
+        this.viewDims.push(i);
+      }
+    }
+
+    if (this.viewDims.length > 2) {
+      throw new Error(`Only selections in 1D and 2D are supported.`);
+    }
+
+    this.rowStart = rowStart == null ? 0 : rowStart;
+    this.colStart = colStart == null ? 0 : colStart;
+    this.rowCount = rowCount == null ? 1 : rowCount;
+    this.colCount = colCount == null ? 1 : colCount;
+  }
+
+  private rank(): number {
+    return this.shape.length;
+  }
+
+  /**
+   * Compute whether a given set of indices falls into the selection.
+   *
+   * ... and if so, whether the set of indices belongs to any of the four
+   * edges of the selected region.
+   *
+   * @param indices
+   * @return Cell selection status, if the set of indices falls into the
+   *   selection. `null` otherwise.
+   */
+  public getElementStatus(indices: number[]): CellSelectionStatus[] | null {
```

Comment:
This function is somewhat long and involves keeping track of several numbers.  Is it possible to break it into testable helper functions?

Reply:
This function itself is thoroughly tested. At about 70 lines, the length of the function is okay IMO.